app/deliver_tx.go

```diff
@@ -0,0 +1,23 @@
+package app
+
+import (
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/x/upgrade"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
+	sdkTx, err := app.txConfig.TxDecoder()(req.Tx)
+	if err == nil {
+		if appVersion, ok := upgrade.IsUpgradeMsg(sdkTx.GetMsgs()); ok {
+			if !IsSupported(appVersion) {
+				panic(fmt.Sprintf("network has upgraded to version %d which is not supported by this node. Please upgrade and restart", appVersion))
+			}
+			app.SetProtocolVersion(appVersion)
+			// TODO: we may want to emit an event for this
+			return abci.ResponseDeliverTx{Code: abci.CodeTypeOK}
+		}
+	}
+	return app.BaseApp.DeliverTx(req)
+}
```

Comment:
> since v1 nodes won't be running this, what will they do with this tx?

They will  just error and skip over it. In the following block they will panic (not sure if in cometbft or in app) because of the app version difference

Reply:
[non-blocking] I think de-indenting and following an exit early strategy makes the code easier to follow and easier to test.  Additionally I think it reduces the chances of future changes causing bugs.
```suggestion
func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
	sdkTx, err := app.txConfig.TxDecoder()(req.Tx)
	if err != nil {
	        return app.BaseApp.DeliverTx(req)
	}
	if appVersion, ok := upgrade.IsUpgradeMsg(sdkTx.GetMsgs()); !ok {
	        return app.BaseApp.DeliverTx(req)
	}
	if !IsSupported(appVersion) {
		panic(fmt.Sprintf("network has upgraded to version %d which is not supported by this node. Please upgrade and restart", appVersion))
	}
	app.SetProtocolVersion(appVersion)
	// TODO: we may want to emit an event for this
	return abci.ResponseDeliverTx{Code: abci.CodeTypeOK}
}
```

Now it's super clear that there are 4 test cases that need to be created.
1. `err != nil` from `app.txConfig.TxDecoder()(req.Tx)`
2. `!ok` from `upgrade.IsUpgradeMsg(sdkTx.GetMsgs())`
3. `!IsSupported(appVersion)`
4. happy case